File: tests/src/test/java/com/netflix/msl/util/Base64Test.java
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.msl.util;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.Collection;

import javax.xml.bind.DatatypeConverter;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import com.netflix.msl.util.Base64.Base64Impl;

/**
 * Base64 tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
@RunWith(Parameterized.class)
public class Base64Test {
    /** UTF-8 charset. */
    private static final Charset CHARSET = Charset.forName("utf-8");
    
    /** Binary Base64 example. */
    private static final String BINARY_B64 = "R0lGODlhPQBEAPeoAJosM//AwO/AwHVYZ/z595kzAP/s7P+goOXMv8+fhw/v739/f+8PD98fH/8mJl+fn/9ZWb8/PzWlwv///6wWGbImAPgTEMImIN9gUFCEm/gDALULDN8PAD6atYdCTX9gUNKlj8wZAKUsAOzZz+UMAOsJAP/Z2ccMDA8PD/95eX5NWvsJCOVNQPtfX/8zM8+QePLl38MGBr8JCP+zs9myn/8GBqwpAP/GxgwJCPny78lzYLgjAJ8vAP9fX/+MjMUcAN8zM/9wcM8ZGcATEL+QePdZWf/29uc/P9cmJu9MTDImIN+/r7+/vz8/P8VNQGNugV8AAF9fX8swMNgTAFlDOICAgPNSUnNWSMQ5MBAQEJE3QPIGAM9AQMqGcG9vb6MhJsEdGM8vLx8fH98AANIWAMuQeL8fABkTEPPQ0OM5OSYdGFl5jo+Pj/+pqcsTE78wMFNGQLYmID4dGPvd3UBAQJmTkP+8vH9QUK+vr8ZWSHpzcJMmILdwcLOGcHRQUHxwcK9PT9DQ0O/v70w5MLypoG8wKOuwsP/g4P/Q0IcwKEswKMl8aJ9fX2xjdOtGRs/Pz+Dg4GImIP8gIH0sKEAwKKmTiKZ8aB/f39Wsl+LFt8dgUE9PT5x5aHBwcP+AgP+WltdgYMyZfyywz78AAAAAAAD///8AAP9mZv///wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAEAAKgALAAAAAA9AEQAAAj/AFEJHEiwoMGDCBMqXMiwocAbBww4nEhxoYkUpzJGrMixogkfGUNqlNixJEIDB0SqHGmyJSojM1bKZOmyop0gM3Oe2liTISKMOoPy7GnwY9CjIYcSRYm0aVKSLmE6nfq05QycVLPuhDrxBlCtYJUqNAq2bNWEBj6ZXRuyxZyDRtqwnXvkhACDV+euTeJm1Ki7A73qNWtFiF+/gA95Gly2CJLDhwEHMOUAAuOpLYDEgBxZ4GRTlC1fDnpkM+fOqD6DDj1aZpITp0dtGCDhr+fVuCu3zlg49ijaokTZTo27uG7Gjn2P+hI8+PDPERoUB318bWbfAJ5sUNFcuGRTYUqV/3ogfXp1rWlMc6awJjiAAd2fm4ogXjz56aypOoIde4OE5u/F9x199dlXnnGiHZWEYbGpsAEA3QXYnHwEFliKAgswgJ8LPeiUXGwedCAKABACCN+EA1pYIIYaFlcDhytd51sGAJbo3onOpajiihlO92KHGaUXGwWjUBChjSPiWJuOO/LYIm4v1tXfE6J4gCSJEZ7YgRYUNrkji9P55sF/ogxw5ZkSqIDaZBV6aSGYq/lGZplndkckZ98xoICbTcIJGQAZcNmdmUc210hs35nCyJ58fgmIKX5RQGOZowxaZwYA+JaoKQwswGijBV4C6SiTUmpphMspJx9unX4KaimjDv9aaXOEBteBqmuuxgEHoLX6Kqx+yXqqBANsgCtit4FWQAEkrNbpq7HSOmtwag5w57GrmlJBASEU18ADjUYb3ADTinIttsgSB1oJFfA63bduimuqKB1keqwUhoCSK374wbujvOSu4QG6UvxBRydcpKsav++Ca6G8A6Pr1x2kVMyHwsVxUALDq/krnrhPSOzXG1lUTIoffqGR7Goi2MAxbv6O2kEG56I7CSlRsEFKFVyovDJoIRTg7sugNRDGqCJzJgcKE0ywc0ELm6KBCCJo8DIPFeCWNGcyqNFE06ToAfV0HBRgxsvLThHn1oddQMrXj5DyAQgjEHSAJMWZwS3HPxT/QMbabI/iBCliMLEJKX2EEkomBAUCxRi42VDADxyTYDVogV+wSChqmKxEKCDAYFDFj4OmwbY7bDGdBhtrnTQYOigeChUmc1K3QTnAUfEgGFgAWt88hKA6aCRIXhxnQ1yg3BCayK44EWdkUQcBByEQChFXfCB776aQsG0BIlQgQgE8qO26X1h8cEUep8ngRBnOy74E9QgRgEAC8SvOfQkh7FDBDmS43PmGoIiKUUEGkMEC/PJHgxw0xH74yx/3XnaYRJgMB8obxQW6kL9QYEJ0FIFgByfIL7/IQAlvQwEpnAC7DtLNJCKUoO/w45c44GwCXiAFB/OXAATQryUxdN4LfFiwgjCNYg+kYMIEFkCKDs6PKAIJouyGWMS1FSKJOMRB/BoIxYJIUXFUxNwoIkEKPAgCBZSQHQ1A2EWDfDEUVLyADj5AChSIQW6gu10bE/JG2VnCZGfo4R4d0sdQoBAHhPjhIB94v/wRoRKQWGRHgrhGSQJxCS+0pCZbEhAAOw==";
    
    /** Standard Base64 examples. */
    private static final Object[][] EXAMPLES = {
        { "The long winded author is going for a walk while the light breeze bellows in his ears.".getBytes(CHARSET),
          "VGhlIGxvbmcgd2luZGVkIGF1dGhvciBpcyBnb2luZyBmb3IgYSB3YWxrIHdoaWxlIHRoZSBsaWdodCBicmVlemUgYmVsbG93cyBpbiBoaXMgZWFycy4=" },
        { "Sometimes porcupines need beds to sleep on.".getBytes(CHARSET),
          "U29tZXRpbWVzIHBvcmN1cGluZXMgbmVlZCBiZWRzIHRvIHNsZWVwIG9uLg==" },
        { "Even the restless dreamer enjoys home-cooked foods.".getBytes(CHARSET),
          "RXZlbiB0aGUgcmVzdGxlc3MgZHJlYW1lciBlbmpveXMgaG9tZS1jb29rZWQgZm9vZHMu" },
        // We use DatatypeConverter here knowing BINARY_B64 is valid and that
        // DatatypeConverter is functionally correct.
        { DatatypeConverter.parseBase64Binary(BINARY_B64),
          BINARY_B64 },
    };
    
    /** Invalid Base64 examples. */
    private static final String[] INVALID_EXAMPLES = {
        "AAAAA",
        "AAAAAAA",
        "%$#@=",
        "ZZZZZZZZZZ=",
        "ZZZZZZZZZ==",
        "U29tZXRpbWVzIHBvcmN1cGluZX=gbmVlZCBiZWRzIHRvIHNsZWVwIG9uLg==",
        "RXZlbiB0aGUgcmVzdGxlc3MgZHJ=YW1lciBlbmpveXMgaG9tZS1jb29rZWQgZm9vZHMu",
        "RXZlbiB0aGUgcmVzdGxlc3MgZHJ=Y",
        "VGhlIGxvbmcgd2luZGVkIGF1dGhvciBpcyBnb2luZyBmb3IgY幸B3YWxrIHdoaWxlIHRoZSBsaWdodCBicmVlemUgYmVsbG93cyBpbiBoaXMgZWFycy4=",
    };
    
    @Parameters
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][] {
            { new Base64Jaxb() },
            { new Base64Secure() }
        });
    }
    
    @Rule
    public ExpectedException thrown = ExpectedException.none();
    
    /**
     * @param impl Base64 encode/decode implementation.
     */
    public Base64Test(final Base64Impl impl) {
        Base64.setImpl(impl);
    }
    
    @Test
    public void standard() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final byte[] data = (byte[])example[0];
            final String base64 = (String)example[1];
            
            // Encode/decode.
            final String encoded = Base64.encode(data);
            final byte[] decoded = Base64.decode(base64);
            
            // Validate.
            assertEquals(base64, encoded);
            assertArrayEquals(data, decoded);
        }
    }
    
    @Test
    public void whitespace() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final byte[] data = (byte[])example[0];
            final String base64 = (String)example[1];
            
            // Modify.
            final int half = base64.length() / 2;
            final String modifiedBase64 = "  \t" + base64.substring(0, half) + "\r\n \r\n\t" + base64.substring(half) + " \t \n";
            
            // Encode/decode.
            final String encoded = Base64.encode(data);
            final byte[] decoded = Base64.decode(modifiedBase64);
            
            // Validate.
            assertEquals(base64, encoded);
            assertArrayEquals(data, decoded);
        }
    }
    
    @Test
    public void invalidPadding() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final String base64 = (String)example[1];
            
            // Modify.
            final String modifiedBase64 = base64 + "=";
            
            // Decode.
            boolean invalid = false;
            try {
                Base64.decode(modifiedBase64);
            } catch (final IllegalArgumentException e) {
                invalid = true;
            }
            assertTrue(invalid);
        }
    }
    
    @Test
    public void injectedPadding() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final String base64 = (String)example[1];
            
            // Modify.
            final int half = base64.length() / 2;
            final String modifiedBase64 = base64.substring(0, half) + "=" + base64.substring(half);
            
            // Decode.
            boolean invalid = false;
            try {
                Base64.decode(modifiedBase64);
            } catch (final IllegalArgumentException e) {
                invalid = true;
            }
            assertTrue(invalid);
        }
    }
    
    @Test
    public void invalidCharacter() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final String base64 = (String)example[1];
            
            // Modify.
            final int half = base64.length() / 2;
            final String modifiedBase64 = base64.substring(0, half) + "|" + base64.substring(half);
            
            // Decode.
            boolean invalid = false;
            try {
                Base64.decode(modifiedBase64);
            } catch (final IllegalArgumentException e) {
                invalid = true;
            }
            assertTrue(invalid);
        }
    }
    
    @Test
    public void outOfRangeCharacter() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final String base64 = (String)example[1];
            
            // Modify.
            final int half = base64.length() / 2;
            final String modifiedBase64 = base64.substring(0, half) + new String(new byte[] {(byte)128}) + base64.substring(half);
            
            // Decode.
            boolean invalid = false;
            try {
                Base64.decode(modifiedBase64);
            } catch (final IllegalArgumentException e) {
                invalid = true;
            }
            assertTrue(invalid);
        }
    }
    
    @Test
    public void invalidLength() {
        for (int i = 0; i < EXAMPLES.length; ++i) {
            // Prepare.
            final Object[] example = EXAMPLES[i];
            final String base64 = (String)example[1];
            
            // Modify.
            final String modifiedBase64 = base64.substring(1);
            
            // Decode.
            boolean invalid = false;
            try {
                Base64.decode(modifiedBase64);
            } catch (final IllegalArgumentException e) {
                invalid = true;
            }
            assertTrue(invalid);
        }
    }
    
    @Test
    public void invalid() {
        for (int i = 0; i < INVALID_EXAMPLES.length; ++i) {
            final String base64 = INVALID_EXAMPLES[i];
            boolean invalid = false;
            try {
                Base64.decode(base64);
            } catch (final IllegalArgumentException e) {
                invalid = true;
            }
            assertTrue(invalid);
        }
    }
    
    @Test
    public void emptyString() {
        final String base64 = "";
        final byte[] b = Base64.decode(base64);
        assertEquals(0, b.length);
    }
}

File: tests/src/test/cpp/util/Base64Test.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <util/Base64.h>
#include <IllegalArgumentException.h>

namespace netflix {
namespace msl {
namespace util {

class Base64Test : public ::testing::Test
{
protected:
    /** Standard Base64 examples. */
    static const int NEXAMPLES = 3;
    const std::string EXAMPLE_STR[NEXAMPLES] = {
        "The long winded author is going for a walk while the light breeze bellows in his ears.",
        "Sometimes porcupines need beds to sleep on.",
        "Even the restless dreamer enjoys home-cooked foods."
    };
    const std::string EXAMPLE_B64[NEXAMPLES] = {
          "VGhlIGxvbmcgd2luZGVkIGF1dGhvciBpcyBnb2luZyBmb3IgYSB3YWxrIHdoaWxlIHRoZSBsaWdodCBicmVlemUgYmVsbG93cyBpbiBoaXMgZWFycy4=",
          "U29tZXRpbWVzIHBvcmN1cGluZXMgbmVlZCBiZWRzIHRvIHNsZWVwIG9uLg==",
          "RXZlbiB0aGUgcmVzdGxlc3MgZHJlYW1lciBlbmpveXMgaG9tZS1jb29rZWQgZm9vZHMu"
    };
    static const int NINVALID_EXAMPLES = 9;
    const std::string INVALID_EXAMPLE_B64[NINVALID_EXAMPLES] = {
            "AAAAA",
            "AAAAAAA",
            "%$#@=",
            "ZZZZZZZZZZ=",
            "ZZZZZZZZZ==",
            "U29tZXRpbWVzIHBvcmN1cGluZX=gbmVlZCBiZWRzIHRvIHNsZWVwIG9uLg==",
            "RXZlbiB0aGUgcmVzdGxlc3MgZHJ=YW1lciBlbmpveXMgaG9tZS1jb29rZWQgZm9vZHMu",
            "RXZlbiB0aGUgcmVzdGxlc3MgZHJ=Y",
            "VGhlIGxvbmcgd2luZGVkIGF1dGhvciBpcyBnb2luZyBmb3IgY幸B3YWxrIHdoaWxlIHRoZSBsaWdodCBicmVlemUgYmVsbG93cyBpbiBoaXMgZWFycy4=",
    };
};

namespace
{

inline std::shared_ptr<std::vector<std::uint8_t>> makevec(const std::string& s)
{
    return std::make_shared<std::vector<std::uint8_t>>(s.begin(), s.end());
}

inline size_t base64Len(size_t inlen) { return ((4ul * inlen / 3ul) + 3ul) & ~3ul; }

} // anonymous namespace

TEST_F(Base64Test, Encode)
{
    // https://tools.ietf.org/html/rfc4648#page-12
    // BASE64("") = ""
    // BASE64("f") = "Zg=="
    // BASE64("fo") = "Zm8="
    // BASE64("foo") = "Zm9v"
    // BASE64("foob") = "Zm9vYg=="
    // BASE64("fooba") = "Zm9vYmE="
    // BASE64("foobar") = "Zm9vYmFy"

    std::shared_ptr<std::string> s;
    std::shared_ptr<std::vector<uint8_t>> v;

    v = makevec("");
    s = Base64::encode(v);
    EXPECT_EQ("", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));

    v = makevec("f");
    s = Base64::encode(v);
    EXPECT_EQ("Zg==", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));

    v = makevec("fo");
    s = Base64::encode(v);
    EXPECT_EQ("Zm8=", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));

    v = makevec("foo");
    s = Base64::encode(v);
    EXPECT_EQ("Zm9v", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));

    v = makevec("foob");
    s = Base64::encode(v);
    EXPECT_EQ("Zm9vYg==", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));

    v = makevec("fooba");
    s = Base64::encode(v);
    EXPECT_EQ("Zm9vYmE=", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));

    v = makevec("foobar");
    s = Base64::encode(v);
    EXPECT_EQ("Zm9vYmFy", *s);
    EXPECT_EQ(s->size(), base64Len(v->size()));
}

TEST_F(Base64Test, Decode)
{
    EXPECT_EQ(*makevec(""), *Base64::decode(std::make_shared<std::string>("")));
    EXPECT_EQ(*makevec("f"), *Base64::decode(std::make_shared<std::string>("Zg==")));
    EXPECT_EQ(*makevec("fo"), *Base64::decode(std::make_shared<std::string>("Zm8=")));
    EXPECT_EQ(*makevec("foo"), *Base64::decode(std::make_shared<std::string>("Zm9v")));
    EXPECT_EQ(*makevec("foob"), *Base64::decode(std::make_shared<std::string>("Zm9vYg==")));
    EXPECT_EQ(*makevec("fooba"), *Base64::decode(std::make_shared<std::string>("Zm9vYmE=")));
    EXPECT_EQ(*makevec("foobar"), *Base64::decode(std::make_shared<std::string>("Zm9vYmFy")));
}

TEST_F(Base64Test, Standard)
{
    for (int i = 0; i < NEXAMPLES; ++i)
    {
        // Prepare.
        std::shared_ptr<ByteArray> data = makevec(EXAMPLE_STR[i]);
        std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

        // Encode/Decode.
        std::shared_ptr<std::string> encoded = Base64::encode(data);
        std::shared_ptr<ByteArray> decoded = Base64::decode(base64);

        // Validate.
        EXPECT_EQ(*base64, *encoded) << "example " << i;
        EXPECT_EQ(*data, *decoded) << "example " << i;
    }
}

TEST_F(Base64Test, Whitespace)
{
    for (int i = 0; i < NEXAMPLES; ++i)
    {
        // Prepare.
        std::shared_ptr<ByteArray> data = makevec(EXAMPLE_STR[i]);
        std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

        // Modify.
        const size_t half = base64->length() / 2;
        std::shared_ptr<std::string> modifiedBase64 = std::make_shared<std::string>("  \t" + base64->substr(0, half) + "\r\n \r\n\t" + base64->substr(half) + " \t \n");

        // Encode/Decode.
        std::shared_ptr<std::string> encoded = Base64::encode(data);
        std::shared_ptr<ByteArray> decoded = Base64::decode(modifiedBase64);

        // Validate.
        EXPECT_EQ(*base64, *encoded) << "example " << i;
        EXPECT_EQ(*data, *decoded) << "example " << i;
    }
}

TEST_F(Base64Test, InvalidPadding)
{
	 for (int i = 0; i < NEXAMPLES; ++i)
	 {
		 // Prepare.
		 std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

		 // Modify.
		 std::shared_ptr<std::string> modifiedBase64 = std::make_shared<std::string>(*base64 + "=");

		 // Decode.
		 bool invalid = false;
		 try {
			 Base64::decode(modifiedBase64);
		 } catch (const IllegalArgumentException&) {
			 invalid = true;
		 }
		 EXPECT_TRUE(invalid);
	 }
}

TEST_F(Base64Test, InjectedPadding)
{
	 for (int i = 0; i < NEXAMPLES; ++i)
	 {
		 // Prepare.
		 std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

		 // Modify.
		 const size_t half = base64->length() / 2;
		 std::shared_ptr<std::string> modifiedBase64 = std::make_shared<std::string>(base64->substr(0, half) + "=" + base64->substr(half));

		 // Decode.
		 bool invalid = false;
		 try {
			 Base64::decode(modifiedBase64);
		 } catch (const IllegalArgumentException&) {
			 invalid = true;
		 }
		 EXPECT_TRUE(invalid);
	 }
}

TEST_F(Base64Test, InvalidCharacter)
{
	 for (int i = 0; i < NEXAMPLES; ++i)
	 {
		 // Prepare.
		 std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

		 // Modify.
		 const size_t half = base64->length() / 2;
		 std::shared_ptr<std::string> modifiedBase64 = std::make_shared<std::string>(base64->substr(0, half) + "|" + base64->substr(half));

		 // Decode.
		 bool invalid = false;
		 try {
			 Base64::decode(modifiedBase64);
		 } catch (const IllegalArgumentException&) {
			 invalid = true;
		 }
		 EXPECT_TRUE(invalid);
	 }
}

TEST_F(Base64Test, OutOfRangeCharacter)
{
	 for (int i = 0; i < NEXAMPLES; ++i)
	 {
		 // Prepare.
		 std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

		 // Modify.
		 const size_t half = base64->length() / 2;
		 std::shared_ptr<std::string> modifiedBase64 = std::make_shared<std::string>(base64->substr(0, half) + static_cast<char>(128) + base64->substr(half));

		 // Decode.
		 bool invalid = false;
		 try {
			 Base64::decode(modifiedBase64);
		 } catch (const IllegalArgumentException&) {
			 invalid = true;
		 }
		 EXPECT_TRUE(invalid);
	 }
}

TEST_F(Base64Test, InvalidLength)
{
	 for (int i = 0; i < NEXAMPLES; ++i)
	 {
		 // Prepare.
		 std::shared_ptr<std::string> base64 = std::make_shared<std::string>(EXAMPLE_B64[i]);

		 // Modify.
		 std::shared_ptr<std::string> modifiedBase64 = std::make_shared<std::string>(base64->substr(1));

		 // Decode.
		 bool invalid = false;
		 try {
			 Base64::decode(modifiedBase64);
		 } catch (const IllegalArgumentException&) {
			 invalid = true;
		 }
		 EXPECT_TRUE(invalid);
	 }
}

TEST_F(Base64Test, Invalid)
{
	 for (int i = 0; i < NINVALID_EXAMPLES; ++i)
	 {
		 std::shared_ptr<std::string> base64 = std::make_shared<std::string>(INVALID_EXAMPLE_B64[i]);
		 bool invalid = false;
		 try {
			 Base64::decode(base64);
		 } catch (const IllegalArgumentException&) {
			 invalid = true;
		 }
		 EXPECT_TRUE(invalid);
	 }
}

TEST_F(Base64Test, RoundTrip)
{
    typedef std::vector<uint8_t> ByteArray;
    const std::string o = // original text
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut eu egestas"
        " nibh, vitae tristique ex. Cras et dui dolor. Cras auctor libero tellu"
        "s, porttitor cursus tellus pulvinar ac. Aenean suscipit mi libero, vit"
        "ae suscipit lorem facilisis et. Quisque gravida leo sapien, sit amet f"
        "inibus augue interdum ac. Vestibulum tristique sed arcu ac aliquet. Cu"
        "m sociis natoque penatibus et magnis dis parturient montes, nascetur r"
        "idiculus mus. Nunc eget malesuada risus";
    // o base64-encoded offline
    std::shared_ptr<std::string> s = std::make_shared<std::string>(
        "TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZW"
        "xpdC4gVXQgZXUgZWdlc3RhcyBuaWJoLCB2aXRhZSB0cmlzdGlxdWUgZXguIENyYXMgZXQg"
        "ZHVpIGRvbG9yLiBDcmFzIGF1Y3RvciBsaWJlcm8gdGVsbHVzLCBwb3J0dGl0b3IgY3Vyc3"
        "VzIHRlbGx1cyBwdWx2aW5hciBhYy4gQWVuZWFuIHN1c2NpcGl0IG1pIGxpYmVybywgdml0"
        "YWUgc3VzY2lwaXQgbG9yZW0gZmFjaWxpc2lzIGV0LiBRdWlzcXVlIGdyYXZpZGEgbGVvIH"
        "NhcGllbiwgc2l0IGFtZXQgZmluaWJ1cyBhdWd1ZSBpbnRlcmR1bSBhYy4gVmVzdGlidWx1"
        "bSB0cmlzdGlxdWUgc2VkIGFyY3UgYWMgYWxpcXVldC4gQ3VtIHNvY2lpcyBuYXRvcXVlIH"
        "BlbmF0aWJ1cyBldCBtYWduaXMgZGlzIHBhcnR1cmllbnQgbW9udGVzLCBuYXNjZXR1ciBy"
        "aWRpY3VsdXMgbXVzLiBOdW5jIGVnZXQgbWFsZXN1YWRhIHJpc3Vz");
    std::shared_ptr<ByteArray> d = util::Base64::decode(s);
    std::shared_ptr<std::string> s_ = util::Base64::encode(d);
    EXPECT_EQ(*s, *s_);
    EXPECT_EQ(ByteArray(o.begin(), o.end()), *d);
}

}}} // namespace netflix::msl::util
